src/Functions/XPath/Parsers/ParserXPathQuery.cpp

```diff
@@ -0,0 +1,45 @@
+#include "ParserXPathQuery.h"
+
+#include <Functions/XPath/ASTs/ASTXPathQuery.h>
+
+#include <Functions/XPath/Parsers/ParserXPathAttribute.h>
+#include <Functions/XPath/Parsers/ParserXPathIndexAccess.h>
+#include <Functions/XPath/Parsers/ParserXPathMemberAccess.h>
+#include <Functions/XPath/Parsers/ParserXPathText.h>
+
+namespace DB
+
+{
+
+bool ParserXPathQuery::parseImpl(Pos & pos, ASTPtr & query, Expected & expected)
+{
+    query = std::make_shared<ASTXPathQuery>();
+
+    ParserXPathAttribute parser_xpath_attribute;
+    ParserXPathIndexAccess parser_xpath_index_access;
+    ParserXPathMemberAccess parser_xpath_member_access;
+    ParserXPathText parser_xpath_text;
+
+    ASTPtr root_member_access;
+    if (!parser_xpath_member_access.parse(pos, root_member_access, expected))
+    {
+        return false;
+    }
+    query->children.push_back(root_member_access);
+
+    ASTPtr accessor;
+    while (parser_xpath_text.parse(pos, accessor, expected) || parser_xpath_attribute.parse(pos, accessor, expected)
```

Comment:
In my opinion, `attribute filter` and `index access` should be children of the corresponding `member access` node. Both specify how to select the member. You'll be able to filter out nodes during traversal, and stop collection unnecessary nodes, when the requested index if found.